src/session/chat/history.rs

```diff
@@ -327,21 +328,41 @@ impl History {
 
     fn remove(&self, message_id: i64) {
         let self_ = imp::History::from_instance(self);
-        let index = self_.list.borrow().iter().position(|m| {
-            if let ItemType::Message(message) = m.type_() {
-                if message.id() == message_id {
-                    return true;
-                }
-            }
 
-            false
-        });
+        if let Some(message) = self_.message_map.borrow_mut().remove(&message_id) {
+            let message_ymd = glib::DateTime::from_unix_utc(message.date() as i64)
+                .unwrap()
+                .ymd();
+
+            // Put this in a block, so that we only need to borrow the list once and the runtime
+            // borrow checker does not panic in Self::items_changed when it borrows the list again.
+            let index = {
+                let mut list = self_.list.borrow_mut();
+
+                // The elements in this list are sorted. While the day dividers are sorted only by
+                // their date time, the messages are additionally sorted by their id. We can
+                // exploit this by applying a binary search.
+                let index = list
+                    .binary_search_by(|m| match m.type_() {
+                        ItemType::Message(message) => message.id().cmp(&message_id),
+                        ItemType::DayDivider(date_time) => {
+                            let ordering = date_time.ymd().cmp(&message_ymd);
```

Comment:
glib::DateTime already implements Ord, so let's use that instead of comparing the `ymd`.